File: 03_Sliding_Window/02_longest_string/main.cpp
#include <iostream>
#include <string>
#include <unordered_set>

class Solution {
public:
    int lengthOfLongestSubstring(std::string s) {
        std::unordered_set<char> chars;
        int left = 0, maxLength = 0;

        for (int right = 0; right < s.size(); ++right) {
            while (chars.find(s[right]) != chars.end()) {
                chars.erase(s[left]);
                ++left;
            }
            chars.insert(s[right]);
            maxLength = std::max(maxLength, right - left + 1);
        }

        return maxLength;
    }
};

int main() {
    Solution solution;
    std::string s1 = "zxyzxyz";
    std::string s2 = "xxxx";

    std::cout << "Length of longest substring without duplicate characters in '" << s1 << "': " << solution.lengthOfLongestSubstring(s1) << std::endl;
    std::cout << "Length of longest substring without duplicate characters in '" << s2 << "': " << solution.lengthOfLongestSubstring(s2) << std::endl;

    return 0;
}

File: 03_Sliding_Window/02_longest_string/main.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_set = set()
        left = 0
        max_length = 0
        
        for right in range(len(s)):
            while s[right] in char_set:
                char_set.remove(s[left])
                left += 1
            char_set.add(s[right])
            max_length = max(max_length, right - left + 1)
        
        return max_length

# Example usage
solution = Solution()
s1 = "zxyzxyz"
s2 = "xxxx"

print(f"Length of longest substring without duplicate characters in '{s1}': {solution.lengthOfLongestSubstring(s1)}")
print(f"Length of longest substring without duplicate characters in '{s2}': {solution.lengthOfLongestSubstring(s2)}")
